temporal-sdk/src/main/java/io/temporal/common/converter/SearchAttributesPayloadConverter.java

```
@@ -0,0 +1,93 @@
+ */
+
+package io.temporal.common.converter;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.PropertyAccessor;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
+import com.google.protobuf.ByteString;
+import io.temporal.api.common.v1.Payload;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.Optional;
+
+public final class SearchAttributesPayloadConverter implements PayloadConverter {
+
+  private static ObjectMapper mapper;
+  private static SearchAttributesPayloadConverter converter;
+
+  private SearchAttributesPayloadConverter() {
+    mapper = new ObjectMapper();
+    mapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
+    mapper.registerModule(new JavaTimeModule());
+    mapper.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY);
+  }
+
+  public static SearchAttributesPayloadConverter getInstance() {
+    if (converter == null) {
+      converter = new SearchAttributesPayloadConverter();
+    }
+    return converter;
+  }
+
+  @Override
+  public String getEncodingType() {
+    return EncodingKeys.METADATA_ENCODING_JSON_NAME;
+  }
+
+  @Override
+  public Optional<Payload> toData(Object obj) throws DataConverterException {
+    try {
+      byte[] serialized = mapper.writeValueAsBytes(obj);
+      return Optional.of(
+          Payload.newBuilder()
+              .putMetadata(EncodingKeys.METADATA_ENCODING_KEY, EncodingKeys.METADATA_ENCODING_JSON)
+              .putMetadata(
+                  EncodingKeys.METADATA_TYPE_KEY,
+                  ByteString.copyFromUtf8(
+                      // Handle NPE
+                      SearchAttributesUtil.javaTypeToEncodedType(obj.getClass()).name()))
```

What if we get Unspecified here?

---

We should discuss this: on options is to treat is a string

---

Talked to Alex. Unspecified should throw an error